service/history/workflow/update/update.go

```diff
@@ -563,15 +588,24 @@ func (u *Update) onResponseMsg(
 	u.instrumentation.countResponseMsg()
 	prevState := u.setState(stateProvisionallyCompleted)
 	eventStore.OnAfterCommit(func(context.Context) {
-		if !u.state.Matches(stateSet(stateAccepted | stateProvisionallyCompleted)) {
+		if !u.state.Matches(stateSet(stateProvisionallyCompleted | stateProvisionallyCompletedAfterAccepted)) {
 			return
 		}
-		u.setState(stateCompleted)
+		beforeCommitState := u.setState(stateCompleted)
 		u.outcome.(*future.FutureImpl[*updatepb.Outcome]).Set(res.GetOutcome(), nil)
+		if beforeCommitState == stateProvisionallyCompletedAfterAccepted {
+			// If Update is accepted and completed on the same WFT, then it gets to ProvisionallyCompleted
+			// state from ProvisionallyAccepted state. OnAfterCommit handler of acceptance
+			// message changes state to ProvisionallyCompletedAfterAccepted.
+			// In this state accepted future must be set after outcome future to prevent race condition
+			// with checks in WaitLifecycleStage. It wasn't set in onAcceptanceMsg, and now it is time to
+			// do it. Update skips Accepted state and goes directly to Completed state.
```

Comment:
```suggestion
			// If the Update is accepted *and* completed in the same WFT, then its state is ProvisionallyCompletedAfterAccepted here, set by onAcceptance's OnAfterCommit.
			//
			// Now, to prevent a race condition in WaitLifecycleStage, the accepted future
		        // has not been set by OnAcceptance, as it must be set *after* the outcome future.
		        // 
			// The Accepted state is skipped, and it transitions straight to Completed.
```